yacl/crypto/primitives/psu/krtw19_psu.cc

```diff
@@ -0,0 +1,228 @@
+// Copyright 2024 zhangwfjh
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "yacl/crypto/primitives/psu/krtw19_psu.h"
+
+#include <algorithm>
+#include <array>
+#include <iterator>
+#include <random>
+#include <unordered_set>
+#include <utility>
+
+#include "yacl/utils/serialize.h"
+
+namespace yacl::crypto {
+
+namespace {
+
+// reference: https://eprint.iacr.org/2019/1234.pdf (Figure 2)
+constexpr float ZETA{0.06f};
+constexpr size_t BIN_SIZE{64ul};  // m+1
+constexpr uint128_t BOT{};
+constexpr size_t NUM_BINS_PER_BATCH{16ul};
+constexpr size_t BATCH_SIZE{NUM_BINS_PER_BATCH * BIN_SIZE};
+
+constexpr size_t NUM_BASE_OT{128ul};
+constexpr size_t NUM_INKP_OT{512ul};
+
+static std::random_device rd;
+static std::mt19937 gen(rd());
+
+struct U128Hasher {
+  size_t operator()(const uint128_t& x) const {
```

Comment:
P? `operator()` may be declared `static`, so you can call this as `U128Hasher::operator(x)`. It's okay to use `struct`, but my recommendation is to replace it with a single function, such as `size_t HashToSizeT(uint128_t x)`.

Reply:
[static operator()](https://en.cppreference.com/w/cpp/feature_test#cpp_static_call_operator) is only supported since C++23. The struct here is used as a function object for hasher in `std::unordered_set`. But I could change it to a lambda.